art/estimators/certification/randomized_smoothing/macer/pytorch.py

```diff
@@ -0,0 +1,238 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2023
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+# persons to whom the Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements MACER applied to classifier predictions.
+
+| Paper link: https://arxiv.org/abs/2001.02378
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+from typing import List, Optional, Tuple, Union, TYPE_CHECKING
+
+from tqdm.auto import trange
+import numpy as np
+
+from art.estimators.certification.randomized_smoothing.pytorch import PyTorchRandomizedSmoothing
+from art.utils import check_and_transform_label_format
+
+if TYPE_CHECKING:
+    # pylint: disable=C0412
+    import torch
+    from art.utils import CLIP_VALUES_TYPE, PREPROCESSING_TYPE
+    from art.defences.preprocessor import Preprocessor
+    from art.defences.postprocessor import Postprocessor
+
+logger = logging.getLogger(__name__)
+
+
+class PyTorchMACER(PyTorchRandomizedSmoothing):
+    """
+    Implementation of MACER training, as introduced in Zhai et al. (2020)
+
+    | Paper link: https://arxiv.org/abs/2001.02378
+    """
+
+    estimator_params = PyTorchRandomizedSmoothing.estimator_params + [
+        "beta",
+        "gamma",
+        "lmbda",
+        "gauss_num",
+    ]
+
+    def __init__(
+        self,
+        model: "torch.nn.Module",
+        loss: "torch.nn.modules.loss._Loss",
+        input_shape: Tuple[int, ...],
+        nb_classes: int,
+        optimizer: Optional["torch.optim.Optimizer"] = None,  # type: ignore
```

Comment:
The `type: ignore` is not required. The original `randomized_smoothing/pytorch.py`had this so I assumed it was needed. It can be removed from all variations including the original.